File: daophot/goodpsf.f
      PROGRAM GOODPSF
C
C=====================================================================
C
C This program generates a file with only good PSF candidates.
C
C=====================================================================
C
      INTEGER nmax
      REAL minchi
      PARAMETER(nmax=1000,minchi=0.5)
C
C 'minchi' is the minimum CHI value allowed for a PSF star.
C You may change this value to fit your needs.
C
      INTEGER i,j,k,n,nr,id(nmax),iid(nmax),nstar,npsf,istar,ibadstar
      REAL x(nmax),y(nmax),mag(nmax),err(nmax),sky(nmax)
      REAL chi(nmax)
      CHARACTER flag(nmax)*1,line*85
      CHARACTER*60 lstfile,chifile,outfile
      CHARACTER*69 line1,line2
      
      PRINT *,'File with PSF candiates (.lst)?'
      READ(*,'(A60)') lstfile
      PRINT *,'File with chi values of PSF candidates (.lst.chi)?'
      READ(*,'(A60)') chifile
      PRINT *,'Output file ?'
      READ(*,'(A60)') outfile
C
C=====================================================================
C Count total number of PSF candidates.
C=====================================================================
C
      OPEN(11,FILE=lstfile)
      READ(11,'(A69)') line1
      READ(11,'(A69)') line2
      READ(11,*)
      nstar = 0
      DO i = 1, nmax
         READ(11,*,END=9) id(i),x(i),y(i),mag(i),err(i),sky(i)
         nstar = nstar + 1
      ENDDO
9     CLOSE(11)
C           
C=====================================================================
C Read in chi values and flags from .chi file.
C Stole this part from Stetson's code (~line 580 of psf.f).
C=====================================================================
C
      k = (nstar-1)/5 + 1
      OPEN(12,FILE=chifile)
      DO i = 1, k
         READ(12,'(A85)') line
         DO istar = i, nstar, k
            j = 17*(istar-i)/k + 1
            READ (line(j:j+16),69) iid(istar),chi(istar),flag(istar)
         ENDDO
      ENDDO
      CLOSE(12)
69    FORMAT(I7, F7.3, 1X, A1, 1X)
C           
C=====================================================================
C  Process PSF candidates flags and prune out.
C=====================================================================
C
      npsf = nstar
      OPEN(21,FILE=outfile)
      WRITE(21,'(A69)') line1
      WRITE(21,'(A69)') line2
      WRITE(21,*)
      DO i = 1, nstar
         IF (flag(i).NE.'?' .AND. flag(i).NE.'*' 
     .                      .AND. chi(i).LT.minchi) THEN
            DO j = 1, nstar
               IF (id(j).EQ.iid(i)) THEN
                  WRITE(21,70) id(j),x(j),y(j),mag(j),err(j),sky(j)
                  GOTO 1
               ENDIF
            ENDDO
         ELSE
            npsf = npsf -1
         ENDIF
1     ENDDO
      CLOSE(21)
70    FORMAT(I7,5F9.3)
C
      WRITE(6,71) npsf,nstar
71    FORMAT(/'You have ',I3,' (out of ',I3,') PSF candidates left.'/)
C
      STOP
      END

File: Python_Wrapper/goodpsf.py
from __future__ import division
from lstManip import lstRead, lstWrite
import numpy as np

'''
This is a rewrite of the fortran program goodpsf.f
It eliminates the need to first go through and replace instances of saturated and defective prior to running this code

SYNTAX:
goodpsf.goodpsf(psfFile, chiFile, outFile, optional keywords)

OPTIONAL KEYWORDS
nmax = maximum number of stars. Residual from the fortran code, it doesn't appear here
minchi = the minimum CHI value allowed for a PSF star. Can be changed.
                                                                                     
'''

def goodpsf(*args, **kwargs):

    if len(args) != 3:
        print 'ERROR, input format is goodpsf(psfFile, chiFile, outFile, kwargs)'
        return                                     

    psfFile, chiFile, outFile = args
    nmax = kwargs.get('nmax', 1000)
    minchi = kwargs.get('minchi', 0.5)

    #Read in psfFile (.lst)
    lData, lHeader = lstRead(psfFile)
    id, x, y, mag, err, sky = lData
    nstar = len(id)

    #Read in CHI file (.psf.log)
    cfileIn = open(chifile,'r').read()
    cfile = cfileIn.split('\n')

    #Find the region where the numbers are and pull that out
    start, stop = [e[0] for e in enumerate(cfile) if ('Profile errors:' in e[1]) | ('File with PSF stars and neighbors' in e[1])]
    subFile = cfile[start+1: stop-1]

    iid = []
    chi = []
    flag = []

    for line in subFile:
        sub = line.strip()
        for s in sub:
            if len(sub) > 0: 
                try: #Is this number an integer? That would make it the id
                    itst = int(s)
                    iid.append(itst)
                except ValueError:
                    try: #Is this number a float? That would make it the chi
                        ftst = float(s)
                        chi.append(ftst)
                    except ValueError:
                        ni = len(iid)
                        nc = len(chi)
                        nf = len(flag)
                        if nc < ni: #If we're missing a chi value, then this string must be a crappy chi indicator
                            chi.append(1.000)
                        elif (nc == ni) & (nf < nc): #If we are not missing a chi, then this string must be a flag
                            flag.append(s)
                        else:
                            print 'ERROR'
        nc = len(chi)
        nf = len(flag)
        if nf < nc: #Put in blank flags where none exist
            flag.append('')

    flag = np.array(flat)
    chi = np.array(chi)
    iid = np.array(iid)

    #Trim star list
    goodInd = (flag != '?') & (flag != '*') & (chi < minchi)
    matchInd = [e[0] for e in enumerate(id) if int(e[1]) in iid[goodInd]]

    #Output new .lst file
    out = (id[matchInd],x[matchInd],y[matchInd],mag[matchInd],err[matchInd],sky[matchInd]))
    lstWrite(out, lHeader, outfile)

    

    
